### ISSUE TITLE
Potential bug in function `CxPlatTimeAtOrBefore32`

### ISSUE THREAD

**Issue by Wizmann** [microsoft/msquic#2597]
```cpp
// Returns TRUE if T1 came before T2.
inline
BOOLEAN
CxPlatTimeAtOrBefore32(
    _In_ uint32_t T1,
    _In_ uint32_t T2
    )
{
    return (int32_t)(T1 - T2) <= 0;
}
```

This function has an integer overflow error which could lead to a totally oppsite result.

Say `T1 = 0` and `T2 = 41234567890`.

You can try run the code [here][1].

[1]: https://ideone.com/UrNFNj

**Comment by nibanks** [comment]
Are you saying you expect `41234567890` (essentially `-1`) to be **after** `0`? This code was specifically written to consider `MAX_UINT32/2` after a number to be **before**.

**Comment by nibanks** [comment]
Think of these numbers as not absolutes, but likely the bottom 32-bits of a much larger number. This function is essentially saying any positive signed 32-bit number difference is considered in the future, and any negative difference is in the past. This only makes sense if you assume the different between the two MUST ALWAYS be less than the maximum signed 32-bit integer.

**Comment by Wizmann** [comment]
It means if you're using a uint32 timestamp, you must not initialize it as `0`. Otherwise it's a 50-50 chance whether it works as expected. 

:(

**Comment by nibanks** [comment]
Yes, you should be initializing it to `TimeNow()` essentially. Is there a case where that is not happening?

**Comment by nibanks** [comment]
So can we close this issue @wizmann?

**Comment by Wizmann** [comment]
Sure.
